cpp1st/week02/taehwan/Templates.cpp

```diff
@@ -0,0 +1,284 @@
+template<typename T>
+class TestExample_3 {
+public:
+    TestExample_3() {};
+    T TestFunc() {
+        return pt_Data;
+    }
+
+protected:
+    static T pt_Data;
+};
+
+template<typename T>
+T TestExample_3<T>::pt_Data = 15;
+
+/// Example 4: function template specialization
+template<typename T>
+T Add(T a, T b) { return a + b; }
+
+template <>
+char * Add(char * left, char * right) {
+    int lengthLeft = strlen(left);
+    int lengthRight = strlen(right);
+    char * string = new char[lengthLeft + lengthRight + 1];
+
+    strcpy_s(string, lengthLeft + 1, left);
+    strcpy_s(string + lengthLeft, lengthRight + 1, right);
+
+    return string;
+}
+
+/// Example 5: class template specialization
+template<typename T>
+class TestExample_5 {
+public:
+    TestExample_5(T parameter) : data(parameter) {}
+
+    T GetData() const { return data; }
+    void SetData(T parameter) { data = parameter; }
+private:
+    T data;
+};
+
+template<>
+class TestExample_5<char *> {
+public:
+    TestExample_5(char * parameter) {
+        int length = strlen(parameter);
+        data = new char[length + 1];
+        strcpy_s(data, length, parameter);
+    }
+
+    ~TestExample_5() { delete [] data; }
+    char * GetData() const { return data; }
+
+private:
+    char * data;
+};
+
+/// example: function pointer wrapper with NULL checking
+template<typename T>
+class Test_WrapFuncObj;
+
+template<typename T, typename ... Args>
+class Test_WrapFuncObj<T(Args ...)> {
+    T (*f)(Args ...);
+public:
+    Test_WrapFuncObj(T (*t)(Args ...)) {
+        f = t;
+    }
+
+    T operator()(Args&& ... args) {
+        if(f != nullptr) {
+            return (*f)(std::forward<Args>(args) ...);
+        }
+        else {
+            return T();
+        }
+    }
+};
+
+class TemplateExample {
+public:
+    void GoTest();
+    void Test1();
+    void Test2();
+    void Test3();
+    void Test4();
+    void Test5();
+    void Test();
+private:
+    void Wrapper(void (TemplateExample::*f)());
+
+    void (TemplateExample::*f)();
+    int counter = 0;
+};
+
+void TemplateExample::Test1() {
+    TestExample_1<int> Foo(5);
+    cout << Foo << endl;
+
+    TestExample_1<double> Bar(3.14);
+    cout << Bar << endl;
+
+    TestExample_1<char *> FooBar("Hello world!");
+    cout << FooBar << endl;
+}
+
+void TemplateExample::Test2() {
+    TestExample_2<int> obj1(5);
+
+    obj1[0] = 1;
+    obj1[1] = 2;
+    obj1[2] = 3;
+    obj1[3] = 4;
+    obj1[4] = 5;
+
+    for(int i = 0; i < 5; i++) {
+        cout << obj1[i] << "\t";
+    }
+    cout << endl;
+
+    TestExample_2<int> obj2(3);
+    obj2 = obj1;
+
+    for(int i = 0; i < 5; i++) {
+        cout << obj2[i] << "\t";
+    }
+    cout << endl;
+}
+
+void TemplateExample::Test3() {
+    TestExample_3<double> test;
+    cout << test.TestFunc() << endl;
+}
+
+void TemplateExample::Test4() {
+    int sum_int = Add<int>(3, 4);
+    cout << sum_int << endl;
+
+    char * sum_char = Add<char *>("Hello ", "World!");
```

Comment:
why don't you use smart pointer?

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f26-use-a-unique_ptrt-to-transfer-ownership-where-a-pointer-is-needed
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r11-avoid-calling-new-and-delete-explicitly

Reply:
I will use a smart pointer next time by referring to your links, thanks!